system/src/usb_control_request_channel.cpp

```diff
@@ -0,0 +1,632 @@
+                return false;
+            }
+            memcpy(req->data, PP_STR(SYSTEM_VERSION_STRING), n);
+            req->wLength = n;
+            break;
+        }
+        default:
+            return false; // Unknown request type
+        }
+    } else {
+        // Host-to-device
+        switch (req->wIndex) {
+        default:
+            return false; // Unknown request type
+        }
+    }
+    return true;
+}
+
+void particle::UsbControlRequestChannel::freeActiveRequest(Request* req) {
+    // Update list of active requests
+    if (req->next) {
+        req->next->prev = req->prev;
+    }
+    if (req->prev) {
+        req->prev->next = req->next;
+    } else {
+        activeReqs_ = req->next;
+    }
+    --activeReqCount_;
+    // Free request data
+    if (req->state == RequestState::ALLOC_PENDING || req->state == RequestState::PENDING) { // ALLOC_PENDING, PENDING
+        // Mark this request as completed to make the system thread free the request data
+        req->state = RequestState::DONE;
+    } else { // ALLOC_FAILED, RECV_PAYLOAD, DONE
+        if (req->request_data && (req->flags & RequestFlag::POOLED_REQ_DATA)) {
+            system_pool_free(req->request_data, nullptr);
+            req->request_data = nullptr;
+        }
+        if (req->reply_data && (req->flags & RequestFlag::POOLED_REP_DATA)) {
+            system_pool_free(req->reply_data, nullptr);
+            req->reply_data = nullptr;
+        }
+        if (!req->request_data && !req->reply_data) {
+            system_pool_free(req, nullptr);
+        } else {
+            // Free the request data asynchronously
+            req->task.func = freeRequest;
+            SystemISRTaskQueue.enqueue(&req->task);
+        }
+    }
+}
+
+void particle::UsbControlRequestChannel::freeRequest(Request* req) {
+    freeRequestData(req);
+    freeReplyData(req);
+    system_pool_free(req, nullptr);
+}
+
+void particle::UsbControlRequestChannel::invokeRequestHandler(ISRTaskQueue::Task* isrTask) {
+    const auto task = static_cast<RequestTask*>(isrTask);
+    const auto req = task->req;
+    const auto channel = static_cast<UsbControlRequestChannel*>(req->channel);
+    channel->handler()->processRequest(req, channel);
+}
+
+void particle::UsbControlRequestChannel::allocRequestData(ISRTaskQueue::Task* isrTask) {
+    const auto task = static_cast<RequestTask*>(isrTask);
+    auto req = task->req;
+    req->request_data = (char*)t_malloc(req->request_size); // FIXME: volatile?
+    ATOMIC_BLOCK() {
+        if (req->state == RequestState::ALLOC_PENDING) {
+            if (req->request_data) {
+                req->state = RequestState::RECV_PAYLOAD; // TODO: Start a timer
+            } else {
+                req->state = RequestState::ALLOC_FAILED;
+            }
+            req = nullptr;
+        }
+    }
+    if (req) { // Request has been cancelled
+        t_free(req->request_data);
+        system_pool_free(req, nullptr);
```

Comment:
doesn't this need unlinking from the list?

Reply:
The list of active requests is always managed in the interrupt handlers, so the list item is already unlinked at this point